feat: add library to perform redaction of sensitive information (#1358)
arcjet/arcjet-js: redact-wasm/README.md

<a href="https://arcjet.com" target="_arcjet-home">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://arcjet.com/logo/arcjet-dark-lockup-voyage-horizontal.svg">
    <img src="https://arcjet.com/logo/arcjet-light-lockup-voyage-horizontal.svg" alt="Arcjet Logo" height="128" width="auto">
  </picture>
</a>

# `@arcjet/redact-wasm`

<p>
  <a href="https://www.npmjs.com/package/@arcjet/redact-wasm">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://img.shields.io/npm/v/%40arcjet%2Fredact-wasm?style=flat-square&label=%E2%9C%A6Aj&labelColor=000000&color=5C5866">
      <img alt="npm badge" src="https://img.shields.io/npm/v/%40arcjet%2Fredact-wasm?style=flat-square&label=%E2%9C%A6Aj&labelColor=ECE6F0&color=ECE6F0">
    </picture>
  </a>
</p>

> This is the [Arcjet][arcjet] Redaction library PII detection engine.

Reviewer: ```suggestion
[Arcjet][arcjet] sensitive information redaction detection engine.
```

## Installation

```shell
npm install -S @arcjet/redact-wasm
```

## Implementation

> This package provides pii identification logic implemented as a WebAssembly module which

Reviewer: ```suggestion
This package provides sensitive information identification and redaction logic implemented as a WebAssembly module which
```

will run local analysis on the provided string.

> The [Wasm](./wasm/arcjet_analyze_bindings_redact.component.js) file contains the binary inlined as
> a base64 [Data URL][mdn-data-url] with the `application/wasm` MIME type.

Reviewer: ```suggestion
The generated `_virtual/arcjet_analyze_bindings_redact.component.js` file contains the binary inlined as
a base64 [Data URL][mdn-data-url] with the `application/wasm` MIME type.
```

This was chosen to save on storage space over inlining the file directly as a
Uint8Array, which would take up ~3x the space of the Wasm file. See
[Better Binary Batter: Mixing Base64 and Uint8Array][wasm-base64-blog] for more
details.

It is then decoded into an ArrayBuffer to be used directly via WebAssembly's
`compile()` function in our entry point file.

This is all done to avoid trying to read or bundle the Wasm asset in various
ways based on the platform or bundler a user is targeting. One example being
that Next.js requires special `asyncWebAssembly` webpack config to load our
Wasm file if we don't do this.

In the future, we hope to do away with this workaround when all bundlers
properly support consistent asset bundling techniques.

> ## API
> 
> In progress.

Reviewer: ```suggestion
```
Remove it or write one?

    <img src="https://arcjet.com/logo/arcjet-light-lockup-voyage-horizontal.svg" alt="Arcjet Logo" height="128" width="auto">
  </picture>
</a>

# `@arcjet/analyze`

<p>
  <a href="https://www.npmjs.com/package/@arcjet/analyze">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://img.shields.io/npm/v/%40arcjet%2Fanalyze?style=flat-square&label=%E2%9C%A6Aj&labelColor=000000&color=5C5866">
      <img alt="npm badge" src="https://img.shields.io/npm/v/%40arcjet%2Fanalyze?style=flat-square&label=%E2%9C%A6Aj&labelColor=ECE6F0&color=ECE6F0">
    </picture>
  </a>
</p>

[Arcjet][arcjet] helps developers protect their apps in just a few lines of
code. Implement rate limiting, bot protection, email verification, and defense
against common attacks.

This is the [Arcjet][arcjet] Redaction library PII detection engine.

## Installation

```shell
npm install -S @arcjet/redact-wasm
```

## Example

```ts
import { generateFingerprint, isValidEmail } from "@arcjet/analyze";

const fingerprint = generateFingerprint("127.0.0.1");
console.log("fingerprint: ", fingerprint);

const valid = isValidEmail("hello@example.com");

> console.log("is email valid?", valid);

Reviewer: Wrong example

Author: I just removed the example instead now that it just provides initialisation. Do you think it makes any sense to keep one around?

Reviewer: I think having an example still makes sense, especially showing the "if defined" check, etc.

Author: Added it back!

```

## Implementation

This package provides pii identification logic implemented as a WebAssembly module which
will run local analysis on the provided string.

> The [arcjet.wasm.js](./wasm/arcjet_analyze_bindings_redact.component.js) file contains the binary inlined as

Reviewer: ```suggestion
The [Wasm](./wasm/arcjet_analyze_bindings_redact.component.js) file contains the binary inlined as
```

Reviewer: I actually realized that this section is totally out of date because we use rollup to output in the _virtual directory. Let me think through rewording it.

    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://img.shields.io/npm/v/%40arcjet%2Fredact-wasm?style=flat-square&label=%E2%9C%A6Aj&labelColor=000000&color=5C5866">
      <img alt="npm badge" src="https://img.shields.io/npm/v/%40arcjet%2Fredact-wasm?style=flat-square&label=%E2%9C%A6Aj&labelColor=ECE6F0&color=ECE6F0">
    </picture>
  </a>
</p>

This is the [Arcjet][arcjet] Redaction library PII detection engine.

## Installation

```shell
npm install -S @arcjet/redact-wasm
```

## Example

```ts
import { initializeWasm } from "@arcjet/redact-wasm";
import type { SensitiveInfoEntity } from "@arcjet/redact-wasm";

function noOpDetect(_tokens: string[]): Array<SensitiveInfoEntity | undefined> {
  return [];
}
function noOpReplace(_input: SensitiveInfoEntity): string | undefined {
  return undefined;
}

const wasm = await initializeWasm(noOpDetect, noOpReplace);

// If WebAssembly isn't available in the environment then it will be undefined.
if (typeof wasm !== "undefined") {
  const config = {
    entities: [],
    contextWindowSize: 1,
    skipCustomDetect: true,
    skipCustomRedact: true,
  };

>   wasm.redact("I am a string", config);

Reviewer: ```suggestion
  const entities = wasm.redact("I am a string", config);
  // Do something with entities that should be redacted.
```
